File: src/parser/italianParser.ts
export type CanonicalVerb =
  | "esamina"
  | "guarda"
  | "vai"
  | "segui"
  | "prendi"
  | "parla"
  | "dai"
  | "usa"
  | "accompagna"
  | "indica"
  | "attacca"
  | "fuggi"
  | "inventario"
  | "aspetta"
  | "aiuto";

export type ParserObject = {
  id: string;
  label: string;
  aliases: string[];
};

export type ParserContext = {
  objects: ParserObject[];
};

export type ParsedCommand = {
  verb: CanonicalVerb;
  targetId?: string;
  secondaryTargetId?: string;
  targetText?: string;
  secondaryTargetText?: string;
};

export type ParserResult =
  | { status: "command"; command: ParsedCommand }
  | { status: "ambiguity"; message: string; verb: CanonicalVerb; matches: ParserObject[] }
  | { status: "unknown"; message: string };

const UNKNOWN_MESSAGE = "Non capisco. Prova aiuto per vedere cosa puoi fare.";

const FILLER_WORDS = new Set([
  "il",
  "lo",
  "la",
  "l",
  "i",
  "gli",
  "le",
  "un",
  "uno",
  "una",
  "del",
  "dello",
  "della",
  "dei",
  "degli",
  "delle",
  "al",
  "allo",
  "alla",
  "ai",
  "agli",
  "alle",
  "a",
  "ad",
  "da",
  "dal",
  "dallo",
  "dalla",
  "dai",
  "dagli",
  "dalle",
  "di",
  "nel",
  "nello",
  "nella",
  "nei",
  "negli",
  "nelle",
  "per",
  "verso"
]);

const VERBS = new Map<string, CanonicalVerb>([
  ["esamina", "esamina"],
  ["esamino", "esamina"],
  ["esaminare", "esamina"],
  ["ispeziona", "esamina"],
  ["ispeziono", "esamina"],
  ["ispezionare", "esamina"],
  ["guarda", "guarda"],
  ["guardo", "guarda"],
  ["guardare", "guarda"],
  ["guardati", "guarda"],
  ["osserva", "guarda"],
  ["osservo", "guarda"],
  ["osservare", "guarda"],
  ["vai", "vai"],
  ["vado", "vai"],
  ["andare", "vai"],
  ["cammina", "vai"],
  ["cammino", "vai"],
  ["camminare", "vai"],
  ["raggiungi", "vai"],
  ["raggiungo", "vai"],
  ["raggiungere", "vai"],
  ["entra", "vai"],
  ["entro", "vai"],
  ["entrare", "vai"],
  ["segui", "segui"],
  ["seguo", "segui"],
  ["seguire", "segui"],
  ["insegui", "segui"],
  ["inseguo", "segui"],
  ["inseguire", "segui"],
  ["prendi", "prendi"],
  ["prendo", "prendi"],
  ["prendere", "prendi"],
  ["raccogli", "prendi"],
  ["raccolgo", "prendi"],
  ["raccogliere", "prendi"],
  ["afferra", "prendi"],
  ["afferro", "prendi"],
  ["afferrare", "prendi"],
  ["parla", "parla"],
  ["parlo", "parla"],
  ["parlare", "parla"],
  ["chiedi", "parla"],
  ["chiedo", "parla"],
  ["chiedere", "parla"],
  ["chiama", "parla"],
  ["chiamo", "parla"],
  ["chiamare", "parla"],
  ["di", "parla"],
  ["dico", "parla"],
  ["dire", "parla"],
  ["dai", "dai"],
  ["do", "dai"],
  ["dare", "dai"],
  ["offri", "dai"],
  ["offro", "dai"],
  ["offrire", "dai"],
  ["porgi", "dai"],
  ["porgo", "dai"],
  ["porgere", "dai"],
  ["regala", "dai"],
  ["regalo", "dai"],
  ["regalare", "dai"],
  ["usa", "usa"],
  ["uso", "usa"],
  ["usare", "usa"],
  ["adopera", "usa"],
  ["adopero", "usa"],
  ["adoperare", "usa"],
  ["mangia", "usa"],
  ["mangio", "usa"],
  ["mangiare", "usa"],
  ["accompagna", "accompagna"],
  ["accompagno", "accompagna"],
  ["accompagnare", "accompagna"],
  ["indica", "indica"],
  ["indico", "indica"],
  ["indicare", "indica"],
  ["mostra", "indica"],
  ["mostro", "indica"],
  ["mostrare", "indica"],
  ["attacca", "attacca"],
  ["attacco", "attacca"],
  ["attaccare", "attacca"],
  ["colpisci", "attacca"],
  ["colpisco", "attacca"],
  ["colpire", "attacca"],
  ["ferisci", "attacca"],
  ["ferisco", "attacca"],
  ["ferire", "attacca"],
  ["fuggi", "fuggi"],
  ["fuggo", "fuggi"],
  ["fuggire", "fuggi"],
  ["scappa", "fuggi"],
  ["scappo", "fuggi"],
  ["scappare", "fuggi"],
  ["inventario", "inventario"],
  ["inv", "inventario"],
  ["zaino", "inventario"],
  ["oggetti", "inventario"],
  ["aspetta", "aspetta"],
  ["attendi", "aspetta"],
  ["attendo", "aspetta"],
  ["attendere", "aspetta"],
  ["esita", "aspetta"],
  ["esito", "aspetta"],
  ["esitare", "aspetta"],
  ["aiuto", "aiuto"],
  ["comandi", "aiuto"],
  ["azioni", "aiuto"]
]);

export function parseItalianCommand(input: string, context: ParserContext): ParserResult {
  const tokens = tokenize(input);

  if (tokens.length === 0) {
    return command({ verb: "aspetta" });
  }

  if (tokens[0] === "?") {
    return command({ verb: "aiuto" });
  }

  const verbToken = tokens[0];
  const parsedVerb = VERBS.get(verbToken);

  if (!parsedVerb) {
    return unknown();
  }

  const rest = tokens.slice(1);

  if (
    parsedVerb === "inventario" ||
    parsedVerb === "fuggi" ||
    parsedVerb === "aspetta" ||
    parsedVerb === "aiuto"
  ) {
    return command({ verb: parsedVerb });
  }

  if (parsedVerb === "guarda") {
    const targetWords = dropFillers(rest);

    if (targetWords.length === 0 || isLookAroundTarget(targetWords)) {
      return command({ verb: "guarda" });
    }

    return resolveObjectCommand("esamina", targetWords, context);
  }

  if (parsedVerb === "vai") {
    return command({ verb: "vai", targetText: targetText(dropFillers(rest)) });
  }

  if (parsedVerb === "attacca") {
    const targetWords = dropFillers(rest);

    if (targetWords.length === 0) {
      return {
        status: "unknown",
        message: missingTargetMessage("attacca")
      };
    }

    return command({ verb: "attacca", targetText: targetText(targetWords) });
  }

  if (parsedVerb === "usa") {
    const splitIndex = rest.indexOf("con");
    const firstTarget = splitIndex >= 0 ? rest.slice(0, splitIndex) : rest;
    const secondTarget = splitIndex >= 0 ? rest.slice(splitIndex + 1) : [];
    const primary = resolveObject(dropFillers(firstTarget), context);

    if (primary.status !== "match") {
      return objectResolutionToResult("usa", primary);
    }

    if (secondTarget.length === 0) {
      return command({
        verb: "usa",
        targetId: primary.object.id,
        targetText: targetText(dropFillers(firstTarget))
      });
    }

    const secondary = resolveObject(dropFillers(secondTarget), context);

    if (secondary.status !== "match") {
      return objectResolutionToResult("usa", secondary);
    }

    return command({
      verb: "usa",
      targetId: primary.object.id,
      secondaryTargetId: secondary.object.id,
      targetText: targetText(dropFillers(firstTarget)),
      secondaryTargetText: targetText(dropFillers(secondTarget))
    });
  }

  return resolveObjectCommand(parsedVerb, dropFillers(rest), context);
}

export function normalizeItalianText(input: string): string {
  return input
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .replace(/['']/g, " ")
    .toLocaleLowerCase("it-IT")
    .replace(/[^\p{L}\p{N}]+/gu, " ")
    .replace(/\s+/g, " ")
    .trim();
}

function tokenize(input: string): string[] {
  if (input.trim() === "?") {
    return ["?"];
  }

  const normalized = normalizeItalianText(input);
  return normalized.length > 0 ? normalized.split(" ") : [];
}

function dropFillers(words: string[]): string[] {
  return words.filter((word) => !FILLER_WORDS.has(word));
}

function resolveObjectCommand(
  verb: CanonicalVerb,
  targetWords: string[],
  context: ParserContext
): ParserResult {
  return objectResolutionToResult(verb, resolveObject(targetWords, context));
}

function objectResolutionToResult(
  verb: CanonicalVerb,
  resolution: ObjectResolution
): ParserResult {
  if (resolution.status === "missing") {
    return {
      status: "unknown",
      message: missingTargetMessage(verb)
    };
  }

  if (resolution.status === "ambiguous") {
    return {
      status: "ambiguity",
      verb,
      matches: resolution.matches,
      message: `Quale? ${formatOptions(resolution.matches)}.`
    };
  }

  if (resolution.status === "unknown") {
    return unknown();
  }

  return command({
    verb,
    targetId: resolution.object.id,
    targetText: targetText(resolution.targetWords)
  });
}

type ObjectResolution =
  | { status: "match"; object: ParserObject; targetWords: string[] }
  | { status: "ambiguous"; matches: ParserObject[] }
  | { status: "missing" }
  | { status: "unknown" };

function resolveObject(targetWords: string[], context: ParserContext): ObjectResolution {
  if (targetWords.length === 0) {
    return { status: "missing" };
  }

  const target = targetText(targetWords);
  const matches = context.objects.filter((object) =>
    object.aliases.some((alias) => {
      const normalizedAlias = targetText(dropFillers(tokenize(alias)));
      return normalizedAlias === target || normalizedAlias.startsWith(`${target} `);
    })
  );

  if (matches.length === 0) {
    return { status: "unknown" };
  }

  if (matches.length > 1) {
    return { status: "ambiguous", matches };
  }

  return { status: "match", object: matches[0], targetWords };
}

function command(commandValue: ParsedCommand): ParserResult {
  return { status: "command", command: commandValue };
}

function unknown(): ParserResult {
  return {
    status: "unknown",
    message: UNKNOWN_MESSAGE
  };
}

function isLookAroundTarget(words: string[]): boolean {
  const target = targetText(words);
  return target === "intorno" || target === "attorno";
}

function targetText(words: string[]): string {
  return words.join(" ").trim();
}

function formatOptions(objects: ParserObject[]): string {
  return objects.map((object) => object.label).join(" o ");
}

function missingTargetMessage(verb: CanonicalVerb): string {
  switch (verb) {
    case "esamina":
      return "Che cosa vuoi esaminare?";
    case "prendi":
      return "Che cosa vuoi prendere?";
    case "parla":
      return "Con chi vuoi parlare?";
    case "dai":
      return "Che cosa vuoi dare?";
    case "usa":
      return "Che cosa vuoi usare?";
    case "attacca":
      return "Dove vuoi colpire?";
    case "segui":
      return "Chi vuoi seguire?";
    case "accompagna":
      return "Chi vuoi accompagnare?";
    case "indica":
      return "Che cosa vuoi indicare?";
    default:
      return UNKNOWN_MESSAGE;
  }
}

File: Assets/Scripts/Core/ItalianParser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

namespace IlLungoViaggio.Core
{
    public enum CanonicalVerb
    {
        Esamina,
        Guarda,
        Vai,
        Segui,
        Prendi,
        Parla,
        Dai,
        Usa,
        Accompagna,
        Indica,
        Dormi,
        Attacca,
        Fuggi,
        Inventario,
        Aspetta,
        Aiuto
    }

    public class ParserObject
    {
        public string Id { get; set; }
        public string Label { get; set; }
        public List<string> Aliases { get; set; }

        public ParserObject(string id, string label, List<string> aliases)
        {
            Id = id;
            Label = label;
            Aliases = aliases;
        }
    }

    public class ParserContext
    {
        public List<ParserObject> Objects { get; set; }

        public ParserContext(List<ParserObject> objects)
        {
            Objects = objects;
        }
    }

    public class ParsedCommand
    {
        public CanonicalVerb Verb { get; set; }
        public string TargetId { get; set; }
        public string SecondaryTargetId { get; set; }
        public string TargetText { get; set; }
        public string SecondaryTargetText { get; set; }
    }

    public enum ParserResultStatus
    {
        Command,
        Ambiguity,
        Unknown
    }

    public class ParserResult
    {
        public ParserResultStatus Status { get; set; }
        public ParsedCommand Command { get; set; }
        public string Message { get; set; }
        public CanonicalVerb? AmbiguityVerb { get; set; }
        public List<ParserObject> Matches { get; set; }
    }

    public static class ItalianParser
    {
        private const string UnknownMessage = "Non capisco. Prova aiuto per vedere cosa puoi fare.";

        private static readonly HashSet<string> FillerWords = new HashSet<string>(StringComparer.Ordinal)
        {
            "il", "lo", "la", "l", "i", "gli", "le",
            "un", "uno", "una",
            "del", "dello", "della", "dei", "degli", "delle",
            "al", "allo", "alla", "ai", "agli", "alle",
            "a", "ad",
            "da", "dal", "dallo", "dalla", "dai", "dagli", "dalle",
            "di",
            "nel", "nello", "nella", "nei", "negli", "nelle",
            "per", "verso"
        };

        private static readonly Dictionary<string, CanonicalVerb> Verbs = new Dictionary<string, CanonicalVerb>(StringComparer.Ordinal)
        {
            { "esamina",    CanonicalVerb.Esamina },
            { "esamino",    CanonicalVerb.Esamina },
            { "esaminare",  CanonicalVerb.Esamina },
            { "ispeziona",  CanonicalVerb.Esamina },
            { "ispeziono",  CanonicalVerb.Esamina },
            { "ispezionare",CanonicalVerb.Esamina },
            { "guarda",     CanonicalVerb.Guarda },
            { "guardo",     CanonicalVerb.Guarda },
            { "guardare",   CanonicalVerb.Guarda },
            { "guardati",   CanonicalVerb.Guarda },
            { "osserva",    CanonicalVerb.Guarda },
            { "osservo",    CanonicalVerb.Guarda },
            { "osservare",  CanonicalVerb.Guarda },
            { "vai",        CanonicalVerb.Vai },
            { "vado",       CanonicalVerb.Vai },
            { "andare",     CanonicalVerb.Vai },
            { "cammina",    CanonicalVerb.Vai },
            { "cammino",    CanonicalVerb.Vai },
            { "camminare",  CanonicalVerb.Vai },
            { "raggiungi",  CanonicalVerb.Vai },
            { "raggiungo",  CanonicalVerb.Vai },
            { "raggiungere",CanonicalVerb.Vai },
            { "entra",      CanonicalVerb.Vai },
            { "entro",      CanonicalVerb.Vai },
            { "entrare",    CanonicalVerb.Vai },
            { "segui",      CanonicalVerb.Segui },
            { "seguo",      CanonicalVerb.Segui },
            { "seguire",    CanonicalVerb.Segui },
            { "insegui",    CanonicalVerb.Segui },
            { "inseguo",    CanonicalVerb.Segui },
            { "inseguire",  CanonicalVerb.Segui },
            { "prendi",     CanonicalVerb.Prendi },
            { "prendo",     CanonicalVerb.Prendi },
            { "prendere",   CanonicalVerb.Prendi },
            { "raccogli",   CanonicalVerb.Prendi },
            { "raccolgo",   CanonicalVerb.Prendi },
            { "raccogliere",CanonicalVerb.Prendi },
            { "afferra",    CanonicalVerb.Prendi },
            { "afferro",    CanonicalVerb.Prendi },
            { "afferrare",  CanonicalVerb.Prendi },
            { "parla",      CanonicalVerb.Parla },
            { "parlo",      CanonicalVerb.Parla },
            { "parlare",    CanonicalVerb.Parla },
            { "chiedi",     CanonicalVerb.Parla },
            { "chiedo",     CanonicalVerb.Parla },
            { "chiedere",   CanonicalVerb.Parla },
            { "chiama",     CanonicalVerb.Parla },
            { "chiamo",     CanonicalVerb.Parla },
            { "chiamare",   CanonicalVerb.Parla },
            { "di",         CanonicalVerb.Parla },
            { "dico",       CanonicalVerb.Parla },
            { "dire",       CanonicalVerb.Parla },
            { "dai",        CanonicalVerb.Dai },
            { "do",         CanonicalVerb.Dai },
            { "dare",       CanonicalVerb.Dai },
            { "offri",      CanonicalVerb.Dai },
            { "offro",      CanonicalVerb.Dai },
            { "offrire",    CanonicalVerb.Dai },
            { "porgi",      CanonicalVerb.Dai },
            { "porgo",      CanonicalVerb.Dai },
            { "porgere",    CanonicalVerb.Dai },
            { "regala",     CanonicalVerb.Dai },
            { "regalo",     CanonicalVerb.Dai },
            { "regalare",   CanonicalVerb.Dai },
            { "usa",        CanonicalVerb.Usa },
            { "uso",        CanonicalVerb.Usa },
            { "usare",      CanonicalVerb.Usa },
            { "adopera",    CanonicalVerb.Usa },
            { "adopero",    CanonicalVerb.Usa },
            { "adoperare",  CanonicalVerb.Usa },
            { "mangia",     CanonicalVerb.Usa },
            { "mangio",     CanonicalVerb.Usa },
            { "mangiare",   CanonicalVerb.Usa },
            { "accompagna", CanonicalVerb.Accompagna },
            { "accompagno", CanonicalVerb.Accompagna },
            { "accompagnare",CanonicalVerb.Accompagna },
            { "indica",     CanonicalVerb.Indica },
            { "indico",     CanonicalVerb.Indica },
            { "indicare",   CanonicalVerb.Indica },
            { "mostra",     CanonicalVerb.Indica },
            { "mostro",     CanonicalVerb.Indica },
            { "mostrare",   CanonicalVerb.Indica },
            { "attacca",    CanonicalVerb.Attacca },
            { "attacco",    CanonicalVerb.Attacca },
            { "attaccare",  CanonicalVerb.Attacca },
            { "colpisci",   CanonicalVerb.Attacca },
            { "colpisco",   CanonicalVerb.Attacca },
            { "colpire",    CanonicalVerb.Attacca },
            { "ferisci",    CanonicalVerb.Attacca },
            { "ferisco",    CanonicalVerb.Attacca },
            { "ferire",     CanonicalVerb.Attacca },
            { "fuggi",      CanonicalVerb.Fuggi },
            { "fuggo",      CanonicalVerb.Fuggi },
            { "fuggire",    CanonicalVerb.Fuggi },
            { "scappa",     CanonicalVerb.Fuggi },
            { "scappo",     CanonicalVerb.Fuggi },
            { "scappare",   CanonicalVerb.Fuggi },
            { "inventario", CanonicalVerb.Inventario },
            { "inv",        CanonicalVerb.Inventario },
            { "zaino",      CanonicalVerb.Inventario },
            { "oggetti",    CanonicalVerb.Inventario },
            { "dormi",      CanonicalVerb.Dormi },
            { "dormo",      CanonicalVerb.Dormi },
            { "dormire",    CanonicalVerb.Dormi },
            { "riposa",     CanonicalVerb.Dormi },
            { "riposo",     CanonicalVerb.Dormi },
            { "riposare",   CanonicalVerb.Dormi },
            { "aspetta",    CanonicalVerb.Aspetta },
            { "attendi",    CanonicalVerb.Aspetta },
            { "attendo",    CanonicalVerb.Aspetta },
            { "attendere",  CanonicalVerb.Aspetta },
            { "esita",      CanonicalVerb.Aspetta },
            { "esito",      CanonicalVerb.Aspetta },
            { "esitare",    CanonicalVerb.Aspetta },
            { "aiuto",      CanonicalVerb.Aiuto },
            { "comandi",    CanonicalVerb.Aiuto },
            { "azioni",     CanonicalVerb.Aiuto }
        };

        public static ParserResult ParseItalianCommand(string input, ParserContext context)
        {
            var tokens = Tokenize(input);

            if (tokens.Count == 0)
                return MakeCommand(new ParsedCommand { Verb = CanonicalVerb.Aspetta });

            if (tokens[0] == "?")
                return MakeCommand(new ParsedCommand { Verb = CanonicalVerb.Aiuto });

            var verbToken = tokens[0];
            if (!Verbs.TryGetValue(verbToken, out CanonicalVerb parsedVerb))
                return MakeUnknown();

            var rest = tokens.Skip(1).ToList();

            switch (parsedVerb)
            {
                case CanonicalVerb.Inventario:
                case CanonicalVerb.Fuggi:
                case CanonicalVerb.Dormi:
                case CanonicalVerb.Aspetta:
                case CanonicalVerb.Aiuto:
                    return MakeCommand(new ParsedCommand { Verb = parsedVerb });

                case CanonicalVerb.Guarda:
                {
                    var targetWords = DropFillers(rest);
                    if (targetWords.Count == 0 || IsLookAroundTarget(targetWords))
                        return MakeCommand(new ParsedCommand { Verb = CanonicalVerb.Guarda });
                    return ResolveObjectCommand(CanonicalVerb.Esamina, targetWords, context);
                }

                case CanonicalVerb.Vai:
                    return MakeCommand(new ParsedCommand
                    {
                        Verb = CanonicalVerb.Vai,
                        TargetText = TargetText(DropFillers(rest))
                    });

                case CanonicalVerb.Parla:
                {
                    var targetWords = DropFillers(rest);
                    if (targetWords.Count == 0)
                        return new ParserResult { Status = ParserResultStatus.Unknown, Message = MissingTargetMessage(CanonicalVerb.Parla) };

                    var resolution = ResolveObject(targetWords, context);
                    if (resolution.Status == ObjectResolutionStatus.Match)
                        return MakeCommand(new ParsedCommand
                        {
                            Verb = CanonicalVerb.Parla,
                            TargetId = resolution.Object.Id,
                            TargetText = TargetText(resolution.TargetWords)
                        });
                    if (resolution.Status == ObjectResolutionStatus.Ambiguous)
                        return new ParserResult
                        {
                            Status = ParserResultStatus.Ambiguity,
                            AmbiguityVerb = CanonicalVerb.Parla,
                            Matches = resolution.Matches,
                            Message = "Quale? " + FormatOptions(resolution.Matches) + "."
                        };
                    return MakeCommand(new ParsedCommand { Verb = CanonicalVerb.Parla, TargetText = TargetText(targetWords) });
                }

                case CanonicalVerb.Attacca:
                {
                    var targetWords = DropFillers(rest);
                    if (targetWords.Count == 0)
                        return new ParserResult { Status = ParserResultStatus.Unknown, Message = MissingTargetMessage(CanonicalVerb.Attacca) };
                    return MakeCommand(new ParsedCommand { Verb = CanonicalVerb.Attacca, TargetText = TargetText(targetWords) });
                }

                case CanonicalVerb.Usa:
                {
                    int splitIndex = rest.IndexOf("con");
                    var firstTarget = splitIndex >= 0 ? rest.Take(splitIndex).ToList() : rest;
                    var secondTarget = splitIndex >= 0 ? rest.Skip(splitIndex + 1).ToList() : new List<string>();
                    var primaryWords = DropFillers(firstTarget);
                    var primary = ResolveObject(primaryWords, context);

                    if (primary.Status != ObjectResolutionStatus.Match)
                        return ObjectResolutionToResult(CanonicalVerb.Usa, primary);

                    if (secondTarget.Count == 0)
                        return MakeCommand(new ParsedCommand
                        {
                            Verb = CanonicalVerb.Usa,
                            TargetId = primary.Object.Id,
                            TargetText = TargetText(primaryWords)
                        });

                    var secondaryWords = DropFillers(secondTarget);
                    var secondary = ResolveObject(secondaryWords, context);

                    if (secondary.Status != ObjectResolutionStatus.Match)
                        return ObjectResolutionToResult(CanonicalVerb.Usa, secondary);

                    return MakeCommand(new ParsedCommand
                    {
                        Verb = CanonicalVerb.Usa,
                        TargetId = primary.Object.Id,
                        SecondaryTargetId = secondary.Object.Id,
                        TargetText = TargetText(primaryWords),
                        SecondaryTargetText = TargetText(secondaryWords)
                    });
                }

                default:
                    return ResolveObjectCommand(parsedVerb, DropFillers(rest), context);
            }
        }

        public static string NormalizeItalianText(string input)
        {
            // Decompose accented chars, strip combining marks
            var decomposed = input.Normalize(NormalizationForm.FormD);
            var sb = new StringBuilder(decomposed.Length);
            foreach (char c in decomposed)
            {
                var cat = System.Globalization.CharUnicodeInfo.GetUnicodeCategory(c);
                if (cat != System.Globalization.UnicodeCategory.NonSpacingMark)
                    sb.Append(c);
            }
            var noAccents = sb.ToString();

            // Replace curly apostrophes with space (matches TS: .replace(/['']/g, " "))
            noAccents = noAccents.Replace('‘', ' ').Replace('’', ' ').Replace('\'', ' ');

            // Lowercase
            var lower = noAccents.ToLowerInvariant();

            // Replace non-letter/non-digit sequences with space
            lower = Regex.Replace(lower, @"[^\p{L}\p{N}]+", " ");

            // Collapse whitespace and trim
            lower = Regex.Replace(lower, @"\s+", " ").Trim();

            return lower;
        }

        // ── private helpers ──────────────────────────────────────────────────────

        private static List<string> Tokenize(string input)
        {
            if (input.Trim() == "?")
                return new List<string> { "?" };

            var normalized = NormalizeItalianText(input);
            if (normalized.Length == 0)
                return new List<string>();
            return normalized.Split(' ').ToList();
        }

        private static List<string> DropFillers(List<string> words)
        {
            return words.Where(w => !FillerWords.Contains(w)).ToList();
        }

        private static string TargetText(List<string> words)
        {
            return string.Join(" ", words).Trim();
        }

        private static bool IsLookAroundTarget(List<string> words)
        {
            var t = TargetText(words);
            return t == "intorno" || t == "attorno";
        }

        private static string FormatOptions(List<ParserObject> objects)
        {
            return string.Join(" o ", objects.Select(o => o.Label));
        }

        private static string MissingTargetMessage(CanonicalVerb verb)
        {
            switch (verb)
            {
                case CanonicalVerb.Esamina:    return "Che cosa vuoi esaminare?";
                case CanonicalVerb.Prendi:     return "Che cosa vuoi prendere?";
                case CanonicalVerb.Parla:      return "Con chi vuoi parlare?";
                case CanonicalVerb.Dai:        return "Che cosa vuoi dare?";
                case CanonicalVerb.Usa:        return "Che cosa vuoi usare?";
                case CanonicalVerb.Attacca:    return "Dove vuoi colpire?";
                case CanonicalVerb.Segui:      return "Chi vuoi seguire?";
                case CanonicalVerb.Accompagna: return "Chi vuoi accompagnare?";
                case CanonicalVerb.Indica:     return "Che cosa vuoi indicare?";
                default:                       return UnknownMessage;
            }
        }

        private static ParserResult MakeCommand(ParsedCommand cmd)
        {
            return new ParserResult { Status = ParserResultStatus.Command, Command = cmd };
        }

        private static ParserResult MakeUnknown()
        {
            return new ParserResult { Status = ParserResultStatus.Unknown, Message = UnknownMessage };
        }

        private static ParserResult ResolveObjectCommand(CanonicalVerb verb, List<string> targetWords, ParserContext context)
        {
            return ObjectResolutionToResult(verb, ResolveObject(targetWords, context));
        }

        private static ParserResult ObjectResolutionToResult(CanonicalVerb verb, ObjectResolution resolution)
        {
            switch (resolution.Status)
            {
                case ObjectResolutionStatus.Missing:
                    return new ParserResult { Status = ParserResultStatus.Unknown, Message = MissingTargetMessage(verb) };
                case ObjectResolutionStatus.Ambiguous:
                    return new ParserResult
                    {
                        Status = ParserResultStatus.Ambiguity,
                        AmbiguityVerb = verb,
                        Matches = resolution.Matches,
                        Message = "Quale? " + FormatOptions(resolution.Matches) + "."
                    };
                case ObjectResolutionStatus.Unknown:
                    return MakeUnknown();
                default: // Match
                    return MakeCommand(new ParsedCommand
                    {
                        Verb = verb,
                        TargetId = resolution.Object.Id,
                        TargetText = TargetText(resolution.TargetWords)
                    });
            }
        }

        private static ObjectResolution ResolveObject(List<string> targetWords, ParserContext context)
        {
            if (targetWords.Count == 0)
                return new ObjectResolution { Status = ObjectResolutionStatus.Missing };

            var target = TargetText(targetWords);
            var matches = context.Objects.Where(obj =>
                obj.Aliases.Any(alias =>
                {
                    var normalizedAlias = TargetText(DropFillers(Tokenize(alias)));
                    return normalizedAlias == target || normalizedAlias.StartsWith(target + " ");
                })
            ).ToList();

            if (matches.Count == 0)
                return new ObjectResolution { Status = ObjectResolutionStatus.Unknown };
            if (matches.Count > 1)
                return new ObjectResolution { Status = ObjectResolutionStatus.Ambiguous, Matches = matches };

            return new ObjectResolution { Status = ObjectResolutionStatus.Match, Object = matches[0], TargetWords = targetWords };
        }

        private enum ObjectResolutionStatus { Match, Ambiguous, Missing, Unknown }

        private class ObjectResolution
        {
            public ObjectResolutionStatus Status { get; set; }
            public ParserObject Object { get; set; }
            public List<string> TargetWords { get; set; }
            public List<ParserObject> Matches { get; set; }
        }
    }
}
